# BFS,Greedy,A* search on 8 Puzzle

In [1]:
puzzle = (7,2,4,
          5,0,6,
          8,3,1)

In [2]:
puzzle

(7, 2, 4, 5, 0, 6, 8, 3, 1)

In [3]:
# def p(l):
#     for i in range(0,len(l),3):
#         print(l[i+0],l[i+1],l[i+2])

In [4]:
# P(puzzle)

In [5]:
start = (7,2,4,
          5,0,6,
          8,3,1)
goal = (0,1,2,3,4,5,6,7,8)

In [6]:
def p(l):
    for i in range(0,len(l),3):
        if(l[i+0]==0):
            print(" ",l[i+1],l[i+2])
        elif(l[i+1]==0):
            print(l[i+0]," ",l[i+2])
        elif(l[i+1]==0):
            print(l[i+0],l[i+1])
        else:
            print(l[i+0],l[i+1],l[i+2])

In [7]:
p(goal)

  1 2
3 4 5
6 7 8


In [8]:
def up(l):
    if(0 in l[0:3]):
        return l
    else:
        ar=list(l)
        i = ar.index(0)
        ar[i-3],ar[i]=ar[i],ar[i-3]
        return tuple(ar)

In [9]:
def down(l):
    if(0 in l[6:9]):
        return l
    else:
        ar = list(l)
        i = ar.index(0)
        ar[i+3],ar[i]=ar[i],ar[i+3]
        return tuple(ar)

In [10]:
def left(l):
    if(l[0]==0 or l[3]==0 or l[6]==0):
        return l
    else:
        ar = list(l)
        i = ar.index(0)
        ar[i-1],ar[i]=ar[i],ar[i-1]
        return tuple(ar)

In [11]:
def right(l):
    if(l[2]==0 or l[5]==0 or l[8]==0):
        return l
    else:
        ar = list(l)
        i = ar.index(0)
        ar[i+1],ar[i]=ar[i],ar[i+1]
        return tuple(ar)

In [12]:
def nexts(x):
    temp = []
    if(x!=up(x)):
        temp.append(up(x))
    if(x!=down(x)):
        temp.append(down(x))
    if(x!=left(x)):
        temp.append(left(x))
    if(x!=right(x)):
        temp.append(right(x))
    return temp

In [13]:
from collections import defaultdict
import heapq

class node:
    def __init__(self, state, parent=None, hc=True):
        self.state = state
        self.parent = parent
        self.h = self.manhattan(state)
        #self.h = self.out_of_sequence(state)
        if self.parent == None:
            self.g = 0
        else:
            self.g = parent.g + 1
        if hc == False:
            self.f = self.h
        else:
            self.f = self.g + self.h
            
    def __lt__(self, z):
        if self.f == z.f:
            return self.g < z.g
        return self.f < z.f

    def manhattan(self, st):
        m = 0
        for i in range(9):
            if st[i] == 0:
                continue
            m += abs(st[i] // 3 - i // 3) + abs(st[i] % 3 - i % 3)
        return m
    
    def out_of_sequence(self,st):
        m = 0
        for i in range(9):
            if(i==0 and st[i]!=0):
                m+=1
            elif(i!=8):
                if(st[i]+1 == st[i+1]):
                    continue
            else:
                if(st[i]==8):
                    continue
                m+=2
        return m

class solve:
    def __init__(self):
        self.states = defaultdict(list)

    def addEdge(self, u, v):
        self.states[u].append(v)
        
    def uninformed(self, start, goal):
        explored = set()
        queue = [[start]]
        while queue:
            route = queue.pop(0)
            node = route[-1]

            if node not in explored:
                neighbours = self.states[node]
                for neighbour in neighbours:
                    new_path = list(route)
                    new_path.append(neighbour)
                    queue.append(new_path)
                    if neighbour == goal:
                        return new_path, len(explored)
            explored.add(node)
        return "IMPOSSIBLE", 0
    
    def Greedy_BFS(self, start, goal):

        frontier = []
        explored = set()
        visited = set()

        heapq.heappush(frontier, node(start, hc=False))
        visited.add(start)

        while frontier:
            n = heapq.heappop(frontier)

            if tuple(n.state) in explored:
                continue

            explored.add(n.state)

            if n.state == goal:
                res = n
                l = len(visited)
                return res, l

            for i in nexts(n.state):
                if i not in visited:
                    z = node(i, n, hc=False)
                    heapq.heappush(frontier, z)
                    visited.add(i)
                elif i not in explored:
                    z = node(i, n, hc=False)
                    heapq.heappush(frontier, z)

        return None, None
    
    def A_star(self, start, goal):

        frontier = []
        explored = set()
        visited = set()
        f = 0
        heapq.heappush(frontier, node(start))
        visited.add(start)

        while frontier:
            n = heapq.heappop(frontier)

            if n.state in explored:
                continue

            explored.add(n.state)

            if n.state == goal:
                res = n
                l = len(visited)
                return res, l

            for i in nexts(n.state):
                if i not in visited:
                    z = node(i, n)
                    heapq.heappush(frontier, z)
                    visited.add(i)
                elif i not in explored:
                    z = node(i, n)
                    heapq.heappush(frontier, z)

        return None, None
    
    def find_path(self, res):
        x = list()

        while True:
            x.append(res.state)
            if (res.parent == None):
                break
            res = res.parent
        x.reverse()
        sn = 1
        for i in x:
            print()
            print(str(sn)+")")
            p(i)
            print()
            sn+=1
        return sn-1

In [14]:
start = (7,2,4,
          5,0,6,
          8,3,1)
goal = (0,1,2,3,4,5,6,7,8)

In [15]:
puzzle = solve()

In [16]:
count = 0
q = set()
q.add(start)

while len(q) != 0:
    if(count%50000==0):
        print(count)
    x = q.pop()
    count+=1
    for i in nexts(x):
        puzzle.addEdge(x,i)
        if i not in puzzle.states.keys():
            if i not in q:
                q.add(i)

0
50000
100000
150000


In [17]:
print("\nUninformed Breadth First search")
path, nodes = puzzle.uninformed(start, goal)
if nodes >0:
    print("\n No of nodes Discovered : ",nodes)
    print()
    sn = 1
    for i in path:
        print()
        print(str(sn)+")")
        p(i)
        print()
        sn+=1
    print("No of moves required : ",len(path))
else:
    print(path)


Uninformed Breadth First search

 No of nodes Discovered :  158790


1)
7 2 4
5   6
8 3 1


2)
7 2 4
  5 6
8 3 1


3)
  2 4
7 5 6
8 3 1


4)
2   4
7 5 6
8 3 1


5)
2 5 4
7   6
8 3 1


6)
2 5 4
7 3 6
8   1


7)
2 5 4
7 3 6
  8 1


8)
2 5 4
  3 6
7 8 1


9)
2 5 4
3   6
7 8 1


10)
2 5 4
3 6 0
7 8 1


11)
2 5 0
3 6 4
7 8 1


12)
2   5
3 6 4
7 8 1


13)
  2 5
3 6 4
7 8 1


14)
3 2 5
  6 4
7 8 1


15)
3 2 5
6   4
7 8 1


16)
3 2 5
6 4 0
7 8 1


17)
3 2 5
6 4 1
7 8 0


18)
3 2 5
6 4 1
7   8


19)
3 2 5
6   1
7 4 8


20)
3 2 5
6 1 0
7 4 8


21)
3 2 0
6 1 5
7 4 8


22)
3   2
6 1 5
7 4 8


23)
3 1 2
6   5
7 4 8


24)
3 1 2
6 4 5
7   8


25)
3 1 2
6 4 5
  7 8


26)
3 1 2
  4 5
6 7 8


27)
  1 2
3 4 5
6 7 8

No of moves required :  27


In [18]:
print("\nInformed Greedy Best First Search")
path, nodes = puzzle.Greedy_BFS(start, goal)
print("\nNo of nodes Discovered : ",nodes)
print()
moves = puzzle.find_path(path)
print("No of moves required : ",moves)


Informed Greedy Best First Search

No of nodes Discovered :  617


1)
7 2 4
5   6
8 3 1


2)
7 2 4
5 3 6
8   1


3)
7 2 4
5 3 6
8 1 0


4)
7 2 4
5 3 0
8 1 6


5)
7 2 0
5 3 4
8 1 6


6)
7   2
5 3 4
8 1 6


7)
  7 2
5 3 4
8 1 6


8)
5 7 2
  3 4
8 1 6


9)
5 7 2
3   4
8 1 6


10)
5 7 2
3 1 4
8   6


11)
5 7 2
3 1 4
  8 6


12)
5 7 2
  1 4
3 8 6


13)
  7 2
5 1 4
3 8 6


14)
7   2
5 1 4
3 8 6


15)
7 1 2
5   4
3 8 6


16)
7 1 2
  5 4
3 8 6


17)
7 1 2
3 5 4
  8 6


18)
7 1 2
3 5 4
8   6


19)
7 1 2
3 5 4
8 6 0


20)
7 1 2
3 5 0
8 6 4


21)
7 1 2
3   5
8 6 4


22)
7 1 2
3 6 5
8   4


23)
7 1 2
3 6 5
  8 4


24)
7 1 2
  6 5
3 8 4


25)
7 1 2
6   5
3 8 4


26)
7 1 2
6 5 0
3 8 4


27)
7 1 2
6 5 4
3 8 0


28)
7 1 2
6 5 4
3   8


29)
7 1 2
6 5 4
  3 8


30)
7 1 2
  5 4
6 3 8


31)
  1 2
7 5 4
6 3 8


32)
1   2
7 5 4
6 3 8


33)
1 5 2
7   4
6 3 8


34)
1 5 2
7 3 4
6   8


35)
1 5 2
7 3 4
  6 8


36)
1 5 2
  3 4
7 6 8


37)
1 5 2
3   4
7 6 8


38)
1   2
3 5 4
7 6 8


39)
  1 2
3 5 4
7 6 8


40)
3

In [19]:
print("\nInformed A* Search")
path, nodes = puzzle.A_star(start, goal)
print("\nNo of nodes Discovered : ",nodes)
moves = puzzle.find_path(path)
print("No of moves required : ",moves)


Informed A* Search

No of nodes Discovered :  6172

1)
7 2 4
5   6
8 3 1


2)
7 2 4
  5 6
8 3 1


3)
  2 4
7 5 6
8 3 1


4)
2   4
7 5 6
8 3 1


5)
2 5 4
7   6
8 3 1


6)
2 5 4
7 3 6
8   1


7)
2 5 4
7 3 6
  8 1


8)
2 5 4
  3 6
7 8 1


9)
2 5 4
3   6
7 8 1


10)
2 5 4
3 6 0
7 8 1


11)
2 5 0
3 6 4
7 8 1


12)
2   5
3 6 4
7 8 1


13)
  2 5
3 6 4
7 8 1


14)
3 2 5
  6 4
7 8 1


15)
3 2 5
6   4
7 8 1


16)
3 2 5
6 4 0
7 8 1


17)
3 2 5
6 4 1
7 8 0


18)
3 2 5
6 4 1
7   8


19)
3 2 5
6 4 1
  7 8


20)
3 2 5
  4 1
6 7 8


21)
3 2 5
4   1
6 7 8


22)
3 2 5
4 1 0
6 7 8


23)
3 2 0
4 1 5
6 7 8


24)
3   2
4 1 5
6 7 8


25)
3 1 2
4   5
6 7 8


26)
3 1 2
  4 5
6 7 8


27)
  1 2
3 4 5
6 7 8

No of moves required :  27


#### Using Manhattan Distance
Uninformed : 158790 nodes 27 moves

Greedy     :    617 nodes 57 moves

A* Search  :   6172 nodes 27 moves